router/fastdp.go

```diff
@@ -0,0 +1,1109 @@
+	lock := fastdp.startLock()
+	defer lock.unlock()
+	return fastdp.bridge(routerBridgePortID, key, &lock)
+}
+
+// Ethernet bridge implementation
+
+func (fastdp *FastDatapath) bridge(ingress bridgePortID,
+	key PacketKey, lock *fastDatapathLock) FlowOp {
+	lock.relock()
+	if fastdp.sendToMAC[key.SrcMAC] == nil {
+		// Learn the source MAC
+		fastdp.sendToMAC[key.SrcMAC] = fastdp.sendToPort[ingress]
+		fastdp.seenMACs[key.SrcMAC] = struct{}{}
+	}
+
+	// If we know about the destination MAC, deliver it to the
+	// associated port.
+	if sender := fastdp.sendToMAC[key.DstMAC]; sender != nil {
+		return NewMultiFlowOp(false, odpEthernetFlowKey(key),
+			sender(key, lock))
+	}
+
+	// Otherwise, it might be a real broadcast, or it might
+	// be for a MAC we don't know about yet.  Either way, we'll
+	// broadcast it.
+	mfop := NewMultiFlowOp(false)
+
+	if (key.DstMAC[0] & 1) == 0 {
+		// Not a real broadcast, so don't create a flow rule.
+		// If we did, we'd need to delete the flows every time
+		// we learned a new MAC address, or have a more
+		// complicated selective invalidation scheme.
+		mfop.Add(vetoFlowCreationFlowOp{})
+	} else {
+		// A real broadcast
+		mfop.Add(odpEthernetFlowKey(key))
+	}
+
+	// Send to all ports except the one it came in on. The
+	// sendToPort map is immutable, so it is safe to iterate over
+	// it even though the sender functions can drop the
+	// fastDatapathLock
+	for id, sender := range fastdp.sendToPort {
+		if id != ingress {
+			mfop.Add(sender(key, lock))
+		}
+	}
+
+	return mfop
+}
+
+// Overlay bits
+
+type fastDatapathOverlay struct {
+	*FastDatapath
+}
+
+func (fastdp *FastDatapath) Overlay() Overlay {
+	return fastDatapathOverlay{fastdp}
+}
+
+func (fastdp fastDatapathOverlay) InvalidateRoutes() {
+	log.Debug("InvalidateRoutes")
+	fastdp.lock.Lock()
+	defer fastdp.lock.Unlock()
+	checkWarn(fastdp.deleteFlows())
+}
+
+func (fastdp fastDatapathOverlay) InvalidateShortIDs() {
+	log.Debug("InvalidateShortIDs")
+	fastdp.lock.Lock()
+	defer fastdp.lock.Unlock()
+	checkWarn(fastdp.deleteFlows())
+}
+
+func (fastDatapathOverlay) AddFeaturesTo(features map[string]string) {
+	// Nothing needed.  Fast datapath support is indicated through
+	// OverlaySwitch.
+}
+
+func (fastdp fastDatapathOverlay) StartConsumingPackets(localPeer *Peer,
+	peers *Peers, consumer OverlayConsumer) error {
+	fastdp.lock.Lock()
+	defer fastdp.lock.Unlock()
+
+	if fastdp.overlayConsumer != nil {
+		return fmt.Errorf("FastDatapath already has an OverlayConsumer")
+	}
+
+	fastdp.localPeer = localPeer
+	fastdp.peers = peers
+	fastdp.overlayConsumer = consumer
+	return nil
+}
+
+func (fastdp *FastDatapath) getVxlanVportID(udpPort int) (odp.VportID, error) {
+	fastdp.lock.Lock()
+	defer fastdp.lock.Unlock()
+
+	if vxlanVportID, present := fastdp.vxlanVportIDs[udpPort]; present {
+		return vxlanVportID, nil
+	}
+
+	vxlanVportID, err := fastdp.dp.CreateVport(
+		odp.NewVxlanVportSpec("vxlan", uint16(udpPort)))
```

Comment:
It would be useful if the vport name included the UDP port (e.g. 'vxlan-6784') - without this the output of `odp flow list weave` is ambiguous in the face of mixed router ports in the same network.